books/compute/src/cuda/kernels/layernorm_forward.md

```diff
@@ -1,11 +1,494 @@
-# Kernels for layernorm forward pass
-
 <!-- Header -->
 
+# Kernels for LayerNorm forward pass
+
 {{ #aipr_header }}
 
 <!-- Main Body -->
 
+## Introduction
+
+The Layer Normalization (LayerNorm) operation applies normalization across the
+last D dimensions of the activation tensor as described in
+[this paper](https://arxiv.org/abs/1607.06450). The normalization equation is
+given below:
+
+$$y = \frac{x - \mathbb{E}[x]}{\sqrt{Var[x] + \epsilon}} * \gamma + \beta$$
+
+This book outlines and provides a detailed explanation of a series of CUDA
+kernel implementations of LayerNorm forward pass based on the
+[llm.c](https://github.com/karpathy/llm.c/tree/master/dev/cuda) repository.
+Please refer to the [Layer Normalization book](../../../../fundamentals/src/normalizations/layernorm.md)
+for conceptual understanding and other details about the operation. For the
+purpose of this book, lets implement kernels for LayerNorm in the
+[Transformer](../../../../nlp/src/llms/architecture/transformer.md) architecture
+for language modeling which expects **a tensor of shape $(B, T, C)$ as input**,
+where $B$ is the batch size, $T$ is the sequence length and $C$ is the hidden
+dimension size. LayerNorm is applied to the last dimension. $B = 8$, $T = 1024$
+and $C = 768$ is set for benchmarking.
+
+The following table shows memory bandwidth for each kernel on a **A40 GPU for
+block size 512**. The last column shows improvement **over the first kernel**:
+
+| Kernel # | Bandwidth (GB/s) | Improvement |
+|:---------|-----------------:|:------------|
+| 1 | 41.43 | - |
+| 2 | 201.25 | 4.9x |
+| 3 | 362.10 | 8.7x |
+| 4 | 432.03 | 10.4x |
+| 5 | 538.88 | 13x |
+
+## Kernel 1
+
+The first kernel is a copy of the CPU implementation. It parallelizes
+over the first 2 dimensions, $B$ and $T$, where $N = B*T$.
```

Comment:
Let's replace $ with ` `.
"The first kernel is a copy of the CPU implementation. It parallelizes over the first 2 dimensions, `B` and `T`, where `N = B * T`."